### PR TITLE
Vowpal Wabbit integration

### PR DESCRIPTION
**Pull request by xdssio**
A [Vowpal Wabbit](https://vowpalwabbit.org/index.html) integrations with *veax.ml*.

```
import vaex 
import vaex.ml
 
ds = vaex.ml.datasets.load_iris()
ds['class_'] = ds['class_'] + 1  # VW classification starts from 1
model = vaex.ml.vowpalwabbit.VowpalWabbitModel(
    params={'oaa': '3', 'P': 1, 'enable_logging': True},
    features=['sepal_length', 'sepal_width', 'petal_length', 'petal_width'],
    target='class_',
    passes=5)
model.fit(ds)
ds = model.transform(ds) 
ds
#    sepal_length    sepal_width    petal_length    petal_width    class_    vowpalwabbit_prediction
0    5.9             3.0            4.2             1.5            2         2
1    6.1             3.0            4.6             1.4            2         2
2    6.6             2.9            4.6             1.3            2         2

### CODE REVIEW

**Review by JovanVeljanoski — CHANGES_REQUESTED** [review]
Hi! 

Excellent work! We've been wanting to support VW for some time now :). This is great. 
Some comments below, to put it in tip-top shape! But it is very close!

One comment: should we use  `num_epochs` instead of `passes`. It is more consistent with IncrementalPredictor for sklearn and the wrapper for River.

This is my review of the code, I will try this in action and will add additional feedback if needed! 

Good job!

**Comment by JovanVeljanoski** [comment]
When using the new class and training & predicting/transforming, a lot of logs are printed. There should be an option to suppress  that. Also since the fitting is done in chunks, there should be a progress bar! See `vaex.ml.sklearn.IncrementalPredictor` for an example.

File: packages/vaex-ml/vaex/ml/vowpalwabbit.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        :return copy: A shallow copy of the DataFrame that includes the Vowpal Wabbit prediction as a virtual column.
+        :rtype: DataFrame
+        '''
+        copy = df.copy()
+        lazy_function = copy.add_function('vowpalwabbit_prediction_function', self, unique=True)
+        expression = lazy_function(*self.features)
+        copy.add_virtual_column(self.prediction_name, expression, unique=False)
+        return copy
+
+    def fit(self, df, passes=1, chunk_size=500, partial_fit=False):
```
[the remark this answers is not in the document]
Comment (xdssio) [reply packages/vaex-ml/vaex/ml/vowpalwabbit.py]: Done

File: packages/vaex-ml/vaex/ml/vowpalwabbit.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        """Fit the VowpalWabbitModel to the DataFrame.
+        :param df: A vaex DataFrame containing the features and target on which to train the model.
+        :param int passes: Number of passes over the data
+        :param int chunk_size: Size of chunks to iterate
+        """
+        passes = passes or self.passes
+        params = {k: v for k, v in self.params.items() if v is not None}
```
[the remark this answers is not in the document]
Comment (xdssio) [reply packages/vaex-ml/vaex/ml/vowpalwabbit.py]: ok

File: packages/vaex-ml/vaex/ml/vowpalwabbit.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        for n in range(passes):
+            for _, _, X in df.to_pandas_df(chunk_size=chunk_size):
+                if n > 1:
+                    X = shuffle(X)
+                for ex in DFtoVW.from_colnames(df=X, y=target, x=features).convert_df():
+                    model.learn(ex)
+        self.model = model
+        return self
+
+    def predict(self, df, **kwargs):
```
[the remark this answers is not in the document]
Comment (xdssio) [reply packages/vaex-ml/vaex/ml/vowpalwabbit.py:116]: I implemented the logic in predict and used it in __call__.
For the sake of consistency, I moved it the other way around as you suggested.

**Comment by xdssio** [comment]
> Hi!
> 
> Excellent work! We've been wanting to support VW for some time now :). This is great. Some comments below, to put it in tip-top shape! But it is very close!
> 
> One comment: should we use `num_epochs` instead of `passes`. It is more consistent with IncrementalPredictor for sklearn and the wrapper for River.
> 
> This is my review of the code, I will try this in action and will add additional feedback if needed!
> 
> Good job!

Fixed

**Comment by xdssio** [comment]
> Hi!
> 
> Excellent work! We've been wanting to support VW for some time now :). This is great. Some comments below, to put it in tip-top shape! But it is very close!
> 
> One comment: should we use `num_epochs` instead of `passes`. It is more consistent with IncrementalPredictor for sklearn and the wrapper for River.
> 
> This is my review of the code, I will try this in action and will add additional feedback if needed!
> 
> Good job!

Think it is done.

**Comment by JovanVeljanoski** [comment]
Nice! Can you add VW to the CI scripts (requirements). Right now it is not being tested in CI since the package is not installed (and there is this "import or skip" statement). 

Ideally,  it would be nice if testing of VW does not depend on `sklearn`, or in that case you need to wrap sklearn in "import or skip" too, since sklearn is not a dependency of vaex or vaex-ml. Maybe best to write a simple util that calculates accuracy, and remove sklearn (in the test script).
